rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,647 @@
+     ...
+```
+
+Here, `OpKernelBase` implementation will be minimal:
+
+```cpp
+class OpKernelBase {
+ public:
+  explicit OpKernelBase(OpKernelConstructionInterface* context) {
+  }
+  virtual ~OpKernelBase() {}
+  virtual void Compute(OpKernelContextInterface* context) = 0;
+};
+```
+
+(For details how extending from `OpKernelBase` instead of `OpKernel` would work
+with current TensorFlow runtime see [Appendix 1](#appendix-1))
+
+Corresponding .cc file then registers the kernel using the correct kernel and
+context classes. For example, this is how we register `AddN` kernel with TFRT:
+
+```cpp
+REGISTER_KERNEL_FALLBACK_KERNEL( "AddN", AddNOp<CPUDevice, int32>);
+```
+
+## Calling kernel
+
+We add a new TFRT BEF kernel called `tfrt_fallback.kernel_fallback`. This kernel directly
+calls a TF kernel’s `Compute` method by creating `TFRTOpKernel*` data structures
+that forward to corresponding TFRT concepts. For example, the following code
+accesses an input in `llvm::ArrayRef<tfrt::RCReference<tfrt::AsyncValue>>` which
+we get from TFRT:
+
+```cpp
+const Tensor& TFRTOpKernelContext::input(int index) {
+  return inputs_[index]->get<Tensor>();
+}
+```
+
+Simplified definition of `tfrt_fallback.kernel_fallback`:
+
+```cpp
+// Instantiate a kernel. This would be a TensorFlow kernel converted to inherit
+// from `OpKernelBase` instead of `OpKernel`.
+std::unique_ptr<OpKernelBase> op  =  …;
+
+// Create TFRTOpKernelContext. The variable exec_ctx here is the tfrt::ExecutionContext passed to the kernel handler.
+TFRTOpKernelContext op_kernel_context(inputs, outputs.size(), op_meta, exec_ctx.host());
+
+// Directly invoke the TF kernel's Compute() method.
+op->Compute(&op_kernel_context);
+```
+
+## tfrt\_fallback.kernel\_fallback call structure
+
+We will be using the following conventions (essentially, these are based on
+Runtime Fallback work that will probably have RFC coming soon):
+
+*   Attributes are passed as key-value pairs, where both key and value are
+    represented as strings.
+*   Types have a specific string representation. We are trying to use names
+    consistent with BEF syntax as much as possible (for e.g. `f32` represents
+    `float`).
+*   Inputs and outputs have type `tensorflow::Tensor`. We will provide BEF
+    kernels to construct these from BEF data (for e.g. constant values).
+
+Example of invoking Conv3D kernel:
+
+```
+%tft_c = "tfd.kernel_fallback"(%tft_a, %tft_b) {
+    _op_name = "Conv3D",
+    attr1_name="data_format", attr1_value="string$NDHWC",
+    attr2_name="strides", attr2_value="list(i32)$1,1,1,1,1",
+    attr3_name="dilations", attr3_value="list(i32)$1,1,1,1,1",
+    attr4_name="padding", attr4_value="padding$SAME"}: (!tfd.tensor, !tfd.tensor) -> !tfd.tensor
+```
+
+For example, `dilations` attribute here has a value of `[1, 1, 1, 1, 1]`.
+Note: TFRT orders attributes by name, alphabetically, which is why we use `attrN_value` and `attrN_name` pattern pair.
```

Comment:
@cky9301 @fdxmw Is this true, and is it documented somewhere?

Reply:
Based on discussion at design review. This is true but not documented.